### PR TITLE
prepare for pip install

### PR DESCRIPTION
**Pull request by TaykYoku**
BEGINRELEASENOTES
Preparing a project for installation using `pip install`.

CHANGE: move main source code to src
NEW: add pyproject.toml, setup.py
CHANGE: add project description to setup.cfg
NEW: move tests things to tests
CHANGE: rename README.md to README.rst

ENDRELEASENOTES

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -1,4 +1,62 @@
...
+
+[options.packages.find]
+where=src
+
+[options.extras_require]
+server =
+testing =
+
+[options.entry_points]
+webappdirac =
```
Comment (chrisburr) [inline setup.cfg]:
```suggestion
dirac =
```

File: setup.cfg [hunk 2]
```diff
@@ -1,4 +1,62 @@
-# This file should be used as the pep8/pycodestyle and autopep8 config
+###############################################################################
+# Setuptools
+###############################################################################
+
+[metadata]
+name = WebAppDIRAC
+description = WebAppDIRAC is a prortal for the DIRAC software.
```
Comment (chrisburr) [inline setup.cfg]:
```suggestion
description = WebAppDIRAC is a portal for the DIRAC software.
```

File: setup.cfg [hunk 3]
```diff
@@ -1,4 +1,62 @@
...
+    #autopep8 ==1.3.3
+    coverage
+    requests >=2.9.1
+    six
+    #mock >=1.0.1
+    #pylint >=1.6.5
+    #pytest >=3.6
+    #pytest-cov >=2.2.0
+    #pytest-mock
+    #tornado >=5.0.0
```
Comment (chrisburr) [inline setup.cfg]: I'm not sure if this is correct but you'll need to include the package data:
```suggestion

zip_safe = False
include_package_data = True

[options.package_data]
* =  WebApp/static
```

You can check by running `python setup.py sdist bdist_wheel` and unzipping the files which are generated to make sure everything you expect is included.

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+[build-system]
+requires = ["setuptools>=42", "wheel", "setuptools_scm[toml]>=3.4"]
+build-backend = "setuptools.build_meta"
+
+# Enable setuptools_scm to compute the version number from the most recent tag
+# https://github.com/pypa/setuptools_scm/#pyprojecttoml-usage
```
Comment (chrisburr) [inline pyproject.toml:6]: Is there a reason for disabling `setuptools_scm`?

Comment (TaykYoku) [reply pyproject.toml:6]: not at all, but had to test how it works. `setuptools_scm` collects the version from git which is not supported in pypi therefore it was added the test version in setup  manual

Comment (TaykYoku) [reply pyproject.toml:6]: in theory I should create a tag with the correct version and uncomment it

File: src/WebAppDIRAC/scripts/dirac_webapp_run.py [hunk 5]
```diff
@@ -5,15 +5,17 @@
...
 from DIRAC.FrameworkSystem.Client.Logger import gLogger
 
 from WebAppDIRAC.Core.App import App
 
 __RCSID__ = "$Id$"
 
-if __name__ == "__main__":
 
+@DiracScript()
+def main()
```
Comment (chrisburr) [inline src/WebAppDIRAC/scripts/dirac_webapp_run.py]:
```suggestion
def main():
```

File: setup.cfg (hunk 3 above)
Comment (TaykYoku) [reply setup.cfg]: tested with `include_package_data`, with `options.package_data ` and without, the result is no different. in all cases js, tpl and others are available in the archive

File: pyproject.toml (hunk 4 above)
Comment (TaykYoku) [reply pyproject.toml:6]: fixed

File: setup.cfg [hunk 6]
```diff
@@ -1,4 +1,66 @@
...
+    Topic :: System :: Distributed Computing
+
+[options]
+python_requires = >=3.8
+package_dir=
+    =src
+packages = find:
+
+install_requires =
+    dirac
```
Comment (chrisburr) [inline setup.cfg]: I guess this should have a version associated with it as I guess this tends to be tightly coupled to DIRAC?
```suggestion
    dirac ~=7.3.0a0
```

File: setup.cfg [hunk 7]
```diff
@@ -1,4 +1,66 @@
...
+    diraccfg
+    autopep8 ==1.3.3
+    coverage
+    requests >=2.9.1
+    six
+    mock >=1.0.1
+    pylint >=1.6.5
+    pytest >=3.6
+    pytest-cov >=2.2.0
+    pytest-mock
```
Comment (chrisburr) [inline setup.cfg]: Many of these are test dependencies and `tornado` should be added as it's used directly. I don't think there is anything else but please check:
```suggestion
    requests >=2.9.1
    tornado
    six
```

File: setup.cfg [hunk 8]
```diff
@@ -1,4 +1,66 @@
...
+include_package_data = True
+
+[options.package_data]
+* = WebApp/static
+
+[options.packages.find]
+where=src
+
+[options.extras_require]
+testing =
```
Comment (chrisburr) [inline setup.cfg:45]: Test only dependencies can go here

File: tests/runPylint.sh [hunk 9]
```diff
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 if [[ "${CHECK}" == "pylintPY3K"  ]]; then
-  find WebApp Lib Core scripts -name "*.py" -and -not -name 'pep8_*' -exec pylint -E --rcfile=.pylintrc3k --py3k --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" {} +
+  find src/WebAppDIRAC/* -name "*.py" -and -not -name 'pep8_*' -exec pylint -E --rcfile=tests/.pylintrc3k --py3k --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" {} +
```
Comment (chrisburr) [inline tests/runPylint.sh]: No need for the `*`:
```suggestion
  find src/WebAppDIRAC/ -name "*.py" -and -not -name 'pep8_*' -exec pylint -E --rcfile=tests/.pylintrc3k --py3k --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" {} +
```

File: tests/runPylint.sh [hunk 10]
```diff
@@ -1,7 +1,7 @@
...
 else
-  find WebApp Lib Core scripts -name "*.py" -and -not -name 'pep8_*' -exec pylint -E --rcfile=.pylintrc --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" {} +
+  find src/WebAppDIRAC/* -name "*.py" -and -not -name 'pep8_*' -exec pylint -E --rcfile=.pylintrc --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" {} +
```
Comment (chrisburr) [inline tests/runPylint.sh]: No need for the `*`:
```suggestion
  find src/WebAppDIRAC/ -name "*.py" -and -not -name 'pep8_*' -exec pylint -E --rcfile=.pylintrc --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" {} +
```

File: setup.cfg (hunk 7 above)
Comment (TaykYoku) [reply setup.cfg]: Correct me, IIUC then WebAppDIRAC should be installed with the dependence of the `dirac` package and it has the dependence of `tornado`, do it still needed to specify both there and here?

Comment (chrisburr) [reply setup.cfg]: You should depend on whatever WebAppDIRAC uses directly. The way to think about it is "if DIRAC suddenly stopped using tornado would WebAppDIRAC still need it?" and in this case the answer is yes.

Comment (TaykYoku) [reply setup.cfg]: Thanks, now I understand what you mean

File: src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js [hunk 11]
```diff
@@ -1,611 +1,611 @@
...
-}
-
-function BlurStack()
-{
-	this.r = 0;
-	this.g = 0;
-	this.b = 0;
-	this.a = 0;
-	this.next = null;
+/*
```
Comment (chrisburr) [inline src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: What is the reason for changing these files? As they're vendored dependencies we should probably leave them in their upstream state.

Comment (TaykYoku) [reply src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: these files have been just moved without any change

Comment (chrisburr) [reply src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: They have been changed a lot:

Screenshot 2021-04-22 at 16 38 39

Comment (TaykYoku) [reply src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: so I see, but the content is the same .. I'll try to delete and move them again

Comment (chrisburr) [reply src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: It had been reformatted, I guess accidentally. I can fix it locally using this but I don't have permission to push:
```bash
$ git checkout upstream/integration WebApp/static/core/js/utils/canvg-1.3/
Updated 4 paths from 108e6fa
$ rm -rf src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/
$ git mv WebApp/static/core/js/utils/canvg-1.3/ src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/
$ git commit -m "Restore content of canvg-1.3"
[v4r3_pyPA3 9dc9af1] Restore content of canvg-1.3
 4 files changed, 3759 insertions(+), 3759 deletions(-)
```

Comment (TaykYoku) [reply src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: as far as i understood my git when moving this files removes the CR from the file encoding, I don't know how typical this behavior is, but I'll try to somehow return the CR to keep the originality of the files..

Comment (chrisburr) [reply src/WebAppDIRAC/WebApp/static/core/js/utils/canvg-1.3/StackBlur.js:1]: Yeah file encoding and line endings can be a real pain...

It's fixed in https://github.com/DIRACGrid/WebAppDIRAC/pull/467.

**Pull request merged**